File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    void combinationSumHelper(vector<int>& candidates, int pos, int target, vector<vector<int>> &ans, vector<int> &a) {
        // base cases
        // we'll always reach the end of array by excluding elements
        if(target == 0){
            ans.push_back(a);
            return;
        }
        if(pos == candidates.size()) {
            return;
        }

        // include curr element on this condition
        if(candidates[pos] <= target) {
            // push curr elem to vector for including it
            a.push_back(candidates[pos]);
            // recursive call by including it
            combinationSumHelper(candidates, pos, target-candidates[pos], ans, a);
            // remove curr elem after it has served it purpose 
            a.pop_back();
        }

        // exclude curr element
        combinationSumHelper(candidates, pos+1, target, ans, a);

    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> a;
        combinationSumHelper(candidates, 0, target, ans, a);
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        
        helper(candidates, 0, target, ans, new ArrayList<Integer>());
        
        return ans;
    }
    
    void helper(int[] arr, int i, int target, List<List<Integer>> ans, ArrayList<Integer> aux) {
        if(target == 0) {
            ans.add(aux);
            return;
        }
        if(i == arr.length) return;
        
        // exclude the element at i & just increment i
        helper(arr, i+1, target, ans, new ArrayList<Integer>(aux));
        
        // include the element at i & don't increment i (because it can be used multiple times in later iterations)
        if(arr[i] <= target) {
            aux.add(arr[i]);
            helper(arr, i, target-arr[i], ans, new ArrayList<Integer>(aux));
        }
        
        
    }
}